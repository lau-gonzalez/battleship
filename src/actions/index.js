import * as helpers from '../lib/index.js';
import { gamePieces } from '../constants';
import * as types from '../constants';

//THUNKS

//-----------------------------------------------------------------------------------------------------------//

//PREGAME-PHASE

export const preGameBoard = (row, col) => {
    return (dispatch, getState) => {
        const state = getState()
        const { selectedPiece, selectedPosition, playerBoard, alreadySelectedShips } = state.gameLogic;
        const { playerShipCount } = state.shipsOnBoard;

        if (!selectedPiece || !selectedPosition) {
            alert('Please select a ship and position');
            return;            
        } else if (alreadySelectedShips.indexOf(selectedPiece) > -1) {
            alert(`You have already chosen the ${selectedPiece}. Please choose another`)
            return;
        }else if (!helpers.checkRangeValid(row, col, gamePieces[selectedPiece], selectedPosition, playerBoard)) {
            alert('Invalid spot!');
            return;
        }

        dispatch(incrementShipCount())
        dispatch(addShip(selectedPiece, selectedPosition, row, col))
        dispatch(addShipToAlreadyChosenList(selectedPiece))

        if (playerShipCount === 4) {
            dispatch(changeGamePhase('battle'))
            alert('START!')
        }
    }
}

//END PREGAME

//--------------------------------------------------------------------------------------//

//BATTLE-PHASE

export const battlePhase = (row, col) => {
    return (dispatch, getState) => {
        const state = getState();
        const { enemyBoard, enemyFleet, turn } = state.gameLogic
        const { enemyBoardHitCount } = state.hitCounts

        if (turn !== 'Player') {
            alert('Wait your turn!');
            return;
          }

        dispatch(destroyEnemySpot(row, col, 'enemyBoard', 'enemyFleet'))

        if (enemyBoard[row][col].piece !== 'E') {
            dispatch(isHit(enemyBoard[row][col], enemyFleet, enemyBoardHitCount))
            dispatch(changeTurn(types.ENEMY_NAME))
            setTimeout(() => dispatch(enemyTurn()), 4000);
        } else {
            dispatch(changeTurn(types.ENEMY_NAME))
            setTimeout(() => dispatch(enemyTurn()), 3000);
        }
    }
}

//--------------------------------------------------------------------------------------//

//ENEMY TURN


const enemyTurn = () => {
    return (dispatch, getState) => {
        const state = getState();
        if (state.gamePhase === 'endGame') { return; }

        const { playerBoard, playerFleet } = state.gameLogic;
        const { playerBoardHitCount } = state.hitCounts;
        const { mode, targetDirection, firstSpotHit, lastSpotHit, didComputerHitLastTurn, targetShipHitCount } = state.computerMoveLogic;

        const { row, col, currentTargetDirection } = helpers.decideWhichSpotToHit(playerBoard, mode, firstSpotHit, lastSpotHit, targetDirection, didComputerHitLastTurn)
        dispatch(destroyEnemySpot(row, col, 'playerBoard', 'playerFleet'))
        if (playerBoard[row][col].piece !== 'E') {
            //enemy hits ship on player board
            if (mode === 'target') {
                dispatch(changeTargetShipHitCount(targetShipHitCount + 1))
                dispatch(changeLastSpotHit(row, col))
                dispatch(changeTargetDirection(currentTargetDirection))
                dispatch(changeTurn(types.PLAYER_NAME))
                dispatch(onPlayerSpotIsHit(playerBoard[row][col], playerFleet, playerBoardHitCount))
            } else {
                dispatch(changeComputerMode('target'))
                dispatch(changeTargetShipHitCount(targetShipHitCount + 1))
                dispatch(changeFirstSpotHit(row, col))
                dispatch(changeTurn(types.PLAYER_NAME))
                dispatch(onPlayerSpotIsHit(playerBoard[row][col], playerFleet, playerBoardHitCount))
            }
        } else {
            if (mode === 'target') {
                if (targetShipHitCount > 1) {
                    dispatch(changeTargetDirection(helpers.getOppositeTargetDirection(targetDirection)))
                    dispatch(changeHitLastTurn(false))

                } else {
                    dispatch(changeTargetDirection(helpers.getNextTargetDirection(targetDirection)))
                    dispatch(changeHitLastTurn(false))
                }
            }
            dispatch(changeTurn(types.PLAYER_NAME));
        }
    }
}


const onPlayerSpotIsHit = (spot, fleet, boardHitCount) => {
    return (dispatch) => {
      if (helpers.isShipDestroyed(spot, fleet)) {     
      
          dispatch(destroyShip(spot, 'playerBoard'))
          dispatch(incrementHitCount('playerBoard'))
          dispatch(changeComputerMode('hunt'))
          dispatch(changeTargetShipHitCount(0))
        
        if (boardHitCount === 4) {
          dispatch(changeGamePhase('endGame'));
        }
      } else {
        dispatch(changeHitLastTurn(true));
      }
    }
  }

// END ENEMY TURN

//-------------------------------------------------------------------------------------------//

const isHit = (spot, fleet, boardHitCount) => {
    return (dispatch) => {
        if (helpers.isShipDestroyed(spot, fleet)) {
            dispatch(destroyShip(spot, 'enemyBoard'))
            dispatch(incrementHitCount('enemyBoard'))

            if (boardHitCount === 4) {
                dispatch(changeGamePhase('endGame'))
            }
        }       

    }
}


//END THUNKS

//------------------------------------------------------------------------------------------------------------------------//

//ACTIONS CREATORS


export const addShip = (selectedPiece, selectedPosition, row, col) => {
    if (!selectedPiece || !selectedPosition) return;
    return {
        type: types.ADD_SHIP,
        piece: selectedPiece,
        pos: selectedPosition,
        row,
        col
    }
}

export const incrementShipCount = () => {
    return {
        type: types.INCREMENT_SHIP_COUNT
    }
}

export const incrementHitCount = (boardType) => {
    return {
        type: types.INCREMENT_HIT_COUNT,
        boardType
    }
}

export const changeGamePhase = (phase) => {
    return {
        type: types.CHANGE_GAME_PHASE,
        phase
    }
}

export const selectShip = (piece, position = 'vertical') => {
    if (!piece) return;
    return {
        type: types.SELECT_SHIP,
        piece,
        position
    }
}

export const addShipToAlreadyChosenList = (selectedShip) => {
    return {
      type: types.ADD_SHIP_TO_ALREADY_CHOSEN_LIST,
      selectedShip
    }
  }

export const selectPosition = (position) => {
    if (!position) return;
    return {
        type: types.SELECT_POSITION,
        position
    }
}

export const inputName = (name) => {
    if (!name) return;
    return {
        type: types.INPUT_NAME,
        name
    }
}

export const destroyEnemySpot = (row, col, board, fleet) => {
    return {
        type: types.DESTROY_SPOT,
        row,
        col,
        board,
        fleet
    }
}

export const destroyShip = (ship, board) => {
    return {
        type: types.DESTROY_SHIP,
        ship,
        board
    }
}

export const changeTurn = (turn) => {
    return {
        type: types.CHANGE_TURN,
        turn
    }
}

export const changeComputerMode = (mode) => {
    return {
      type: types.CHANGE_COMPUTER_MODE,
      mode
    }
  }
  
  export const changeTargetDirection = (targetDirection) => {
    return {
      type: types.CHANGE_TARGET_DIRECTION,
      targetDirection
    }
  }
  
  export const changeFirstSpotHit = (row, col) => {
    return {
      type: types.CHANGE_FIRST_SPOT_HIT,
      row,
      col
    }
  }
  export const changeLastSpotHit = (row, col) => {
    return {
      type: types.CHANGE_LAST_SPOT_HIT,
      row, 
      col
    }
  }
  
  export const changeHitLastTurn = (hit) => {
    return {
      type: types.CHANGE_HIT_LAST_TURN,
      hit
    }
  }

  export const changeTargetShipHitCount = (hits) => {
    return {
        type: types.CHANGE_TARGET_SHIP_HIT_COUNT,
        hits
      }
    }