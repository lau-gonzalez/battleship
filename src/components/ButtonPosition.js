import React from 'react'

export default function ButtonShip(props) {
    const selectionClass = props.selectedPosition && props.selectedPosition === props.name ? 'selected' : 'unselected'
    return (
        <div>
            <button className={`btn btn-dark mr-3 ${selectionClass}`}
             name={props.name} onClick={e => {props.onButtonClick(e.target.name)}}>{props.name}</button>
            
        </div>
    )
}
