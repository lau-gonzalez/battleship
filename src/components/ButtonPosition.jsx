/* eslint linebreak-style: ["error", "windows"] */
import React from 'react';
import PropTypes from 'prop-types';

export default function ButtonShip(props) {
  const { name, onButtonClick } = props;
  return (
    <div>
      <button
        type="button"
        className="btn btn-dark mr-3"
        name={name}
        onClick={(e) => { onButtonClick(e.target.name); }}
      >
        {name}
      </button>

    </div>
  );
}

ButtonShip.propTypes = {
  name: PropTypes.string.isRequired,
  onButtonClick: PropTypes.func.isRequired,
};
