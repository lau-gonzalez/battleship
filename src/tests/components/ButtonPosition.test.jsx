/* eslint-disable no-console */
/* eslint-disable no-undef */
/* eslint linebreak-style: ["error", "windows"] */
import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import ButtonPosition from '../../components/ButtonPosition';
import { selectPosition } from '../../actions';

const props = {
  name: 'CR',
  onButtonClick: () => {},
};

function shallowWithRedux(component,
  { initialState, store = createStore(selectPosition, initialState) } = {}) {
  return {
    ...shallow(<Provider store={store}>{component}</Provider>),
  };
}

describe('../../components/Header', () => {
  it('should render correct', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const wrapper = shallowWithRedux(<ButtonPosition {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
