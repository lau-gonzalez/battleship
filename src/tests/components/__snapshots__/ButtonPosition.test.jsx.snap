// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`../../components/Header should render correct 1`] = `Object {}`;
